index.bs

```diff
@@ -6145,6 +6282,358 @@ However, [=authenticators=] that do not utilize [[!FIDO-CTAP]] do not necessaril
+
+        1. Let the values of the `$$attStmtType` "group socket" [[=CDDL=]] be the result of generating an [=attestation statement=] with the [=attestation statement format=], |attFormat|. See [[#sctn-device-publickey-attestation-calculations]].
+
+            Note: The details of the `$$attStmtType` values are dependent upon the particular [=attestation statement=] format. See [[#sctn-attestation-formats]]. Since the [=hardware-bound device key pair=] is specific to a particular authenticator, its attestation can be tied to hardware roots of trust (although they do not have to be). This is in contrast to the associated [=user credential=]'s attestation.
+
+        1. If the `$$attStmtType` "group socket" contains uniquely identifying information then let `epAtt` be [TRUE]. Otherwise omit the `epAtt` field. (This field can only be [TRUE] when the {{AuthenticationExtensionsDevicePublicKeyInputs/attestation}} field of the extension input is "enterprise" and either the user-agent or the authenticator permits uniquely identifying attestation for the requested RP ID.)
+
+    1. Let |dpkSig| be the result of signing the [=assertion signature=] [input](#fig-signature) with |devicePrivateKey|.
+
+    1. Output |dpkSig| as the extension's [=unsigned extension output=].
+
+        Note: |dpkSig| cannot be included in the [=authenticator extension output=] because it is returned inside the [=authenticator data=] and that would imply that the signature signs over itself.
+
+#### Attestation calculations #### {#sctn-device-publickey-attestation-calculations}
+
+When computing attestations, the process in [[#sctn-generating-an-attestation-object]] takes two inputs: `authData` and `hash`. When calculating an attestation for a [=device-bound key=], the typical value for `hash` hashes over the attestation signature itself, which is impossible. Also the attestation of a [=device-bound key=] is potentially used repeatedly, thus may want to be cached. But signing over values that include [=[RP]=]-chosen nonces, like the [=hash of the serialized client data=], make that impossible.
+
+Therefore when calculating an attestation for a [=device-bound key=], the inputs are:
+
+    * For `authData`, substitute the concatenation of the byte string h'64657669636520626f756e64206b6579206174746573746174696f6e2073696700ffffffff' and the value of |aaguid| from the extension output.
+    * For `hash`, substitute the concatenation of the |dpk| and |nonce| fields from the extension output. (The nonce may be empty.)
+
+The attestation signature is thus typically calculated over the bytes of <code>(h'64657669636520626f756e64206b6579206174746573746174696f6e2073696700ffffffff' || |aaguid| || |dpk| || |nonce|)</code>. The 37-byte prefix ensures domain separation: it takes the place of the RP ID hash, flags, and signature counter fields in those messages and ensures that no attestation signature for a [=device-bound key=] can be confused with a signature for a [=user credential=].
```

Comment:
Is it "always unique"? It depends on the challenge being unique, right?

Reply:
Wanted to check if you feel there's still a change to be made here. An attestation can be replayed, but it's still attesting to a true statement and any assertion should also come with a signature from the attested key and that's the thing that's hard to fake. Is it worth writing that in a note, you feel?

Reply:
I saw now that the suggested paragraph above has been added already, and I'm coming back to thinking it (the "Note that when _nonce_ is empty..." paragraph) is probably not necessary after all. It's probably evident enough if you think about it.

For the dpk attestation to be protected from being replayed on a different device, the RP must make sure that the `dpk` the attestation signs over is the same as the key that generated _dpkSig_, right? But it looks like that check is also covered (though perhaps implicitly) in the verification procedures, I think.

Reply:
> the RP must make sure that the dpk the attestation signs over is the same as the key that generated dpkSig, right?

Correct. The attestation says that a given key was created in a certain class of device. That statement can be replayed elsewhere because it's still true. But the RP should only care about that if there's a proof-of-possession of that same key! Otherwise the attestation is making a true statement about something unrelated.

Reply:
Right, thanks.

Because of issue #1799 I'm also swinging back to thinking this kind of clarification is still needed (so I no longer think the "Note that when _nonce_ is empty..." paragraph is unnecessary) - perhaps we need to explain it even more explicitly. But let's cover that in #1799 and resolve this thread.